spec/WebauthnAuthenticationAdapter.spec.js

```diff
@@ -0,0 +1,407 @@
+    expect(webauthnAuthData.id).toEqual(clientRegistration.id);
+    expect(webauthnAuthData.counter).toEqual(0);
+    expect(typeof webauthnAuthData.publicKey).toEqual('string');
+  });
+  it('should login', async () => {
+    const server = await reconfigureServer({
+      auth: { webauthn: { options: { rpId: authenticationRpId, origin: authenticationOrigin } } },
+    });
+    const user = new Parse.User();
+    await user.save({ username: 'username', password: 'password' });
+    await server.config.databaseController.update(
+      '_User',
+      { objectId: user.id },
+      { authData: { webauthn: authenticationCredential } },
+      {}
+    );
+
+    const { signedChallenge, options } = await callChallenge();
+    expect(typeof signedChallenge).toEqual('string');
+    expect(typeof options.challenge).toEqual('string');
+    expect(options.timeout).toEqual(60000);
+
+    const user2 = new Parse.User();
+
+    await expectAsync(
+      user2.save({
+        authData: {
+          webauthn: {
+            id: authenticationCredential.id,
+            // Authentication is missing
+          },
+        },
+      })
+    ).toBeRejectedWithError('authentication is required.');
+
+    await expectAsync(
+      user2.save({
+        authData: {
+          webauthn: {
+            id: authenticationCredential.id,
+            authentication: clientAuthentication,
+            signedChallenge: sign({ challenge: 'test' }, jwtSecret),
+          },
+        },
+      })
+    ).toBeRejectedWithError('Invalid webauthn authentication');
+
+    await user2.save({
+      authData: {
+        webauthn: {
+          id: authenticationCredential.id,
+          authentication: clientAuthentication,
+          signedChallenge: sign({ challenge: authenticationChallenge }, jwtSecret),
+        },
+      },
+    });
+
+    expect(user2.getSessionToken()).toBeDefined();
+    expect(user2.id).toEqual(user.id);
+    await user2.fetch({ useMasterKey: true });
+    const webauthnAuthData = user2.get('authData').webauthn;
+    expect(webauthnAuthData.publicKey).toEqual(authenticationCredential.publicKey);
+    expect(webauthnAuthData.id).toEqual(authenticationCredential.id);
+    expect(webauthnAuthData.counter).toEqual(authenticationCredential.counter);
+  });
+  it('should handle options rpId, rpName, origin, getUsername, getUserDisplayName, attestationType, requireResidentKey', async () => {
+    await reconfigureServer({
+      auth: { webauthn: true },
+    });
+
+    const user = new Parse.User();
+    await user.save({ username: 'username', password: 'password', email: 'test@test.test' });
+
+    let options = (await callChallenge(user.getSessionToken())).options;
+
+    expect(options.rp).toEqual({ name: 'Localhost', id: 'localhost' });
+    expect(options.user).toEqual({ id: user.id, name: 'username', displayName: 'test@test.test' });
+
+    await reconfigureServer({
+      publicServerURL: 'https://subdomain.a-super-app.com/parse',
```

Comment:
Should we enforce `example.parseplatform.org/parse` or something for placeholder urls? cc @mtrezza